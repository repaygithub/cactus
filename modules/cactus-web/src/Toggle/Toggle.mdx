---
name: Toggle
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import Toggle from './Toggle'
import cactusTheme from '@repay/cactus-theme'
import React, { useState } from 'react'

# Toggle

The `Toggle` component can be useful for boolean fields.

### How it loooks

export const Wrapper = () => {
  const [state, setState]=useState(true)
  return <Toggle value={state}  onClick={()=> setState(!state)}/>
}

<Wrapper />


## Best practices

- If you are associating a label to the toggle manually, it's important to connect the label to the toggle using the
  `htmlFor` prop. This will ensure that accessibility tools are able to determine what the toggle is for. For that reason,
  it's also important to use descriptive labels.
  - Example:
  ```jsx
  <Toggle id="toggle" value={false} />
  <Label htmlFor="toggle">The Great War</Label>
  ```

## Basic usage

```jsx
import { Toggle } from '@repay/cactus-web'
...
// Basic usage
<Toggle value={false} />

// Disabled
<Toggle value={true} disabled={true} />
```

## Properties

<PropsTable of={Toggle} />
