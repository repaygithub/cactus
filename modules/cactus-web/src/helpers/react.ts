import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
