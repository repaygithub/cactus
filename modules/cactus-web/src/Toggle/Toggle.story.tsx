import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean, select } from '@storybook/addon-knobs'
import Toggle from './Toggle'

const toggleStories = storiesOf('Toggle', module)

const initialState = { value: false }
type State = Readonly<typeof initialState>

class ToggleManager extends React.Component {
  readonly state: State = initialState
  handleToggle = () => {
    this.setState({
      value: !this.state.value,
    })
  }
  render() {
    return (
      <Toggle
        value={this.state.value}
        onClick={this.handleToggle}
        disabled={boolean('disabled', false)}
        {...this.props}
      />
    )
  }
}

toggleStories.add('Basic Usage', () => {
  return <ToggleManager />
})
