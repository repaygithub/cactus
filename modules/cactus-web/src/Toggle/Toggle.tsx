import React from 'react'

import { margin, MarginProps } from 'styled-system'
import { NavigationClose, StatusCheck } from '@repay/cactus-icons'
import { Omit } from '../types'
import { omitMargins } from '../helpers/omit'
import PropTypes from 'prop-types'
import styled from 'styled-components'

export interface ToggleProps
  extends Omit<
      React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>,
      'value' | 'ref'
    >,
    MarginProps {
  value?: boolean
  disabled?: boolean
}

const ToggleBase = (props: ToggleProps) => {
  const componentProps = omitMargins(props)
  const { value, ...toggleProps } = componentProps
  return (
    <button type="button" role="switch" aria-checked={value} {...toggleProps}>
      <StyledX aria-hidden="true" />
      <StyledCheck aria-hidden="true" />
    </button>
  )
}

const StyledX = styled(NavigationClose)`
  width: 12px;
  height: 12px;
  position: absolute;
  top: 6px;
  right: 8px;
  color: ${p => p.theme.colors.white};
`

const StyledCheck = styled(StatusCheck)`
  width: 15px;
  height: 15px;
  position: absolute;
  top: 4px;
  left: 5px;
  color: ${p => p.theme.colors.white};
`

export const Toggle = styled(ToggleBase)`
  position: relative;
  width: 51px;
  height: 26px;
  border-radius: 13px;
  outline: none;
  background-color: ${p => p.theme.colors.error};
  border: 1px solid ${p => p.theme.colors.error};
  cursor: ${p => (p.disabled ? 'cursor' : 'pointer')};

  &:focus {
    box-shadow: 0 0 8px ${p => p.theme.colors.callToAction};
  }

  ::after {
    width: 26px;
    height: 26px;
    border-radius: 50%;
    content: '';
    top: -1px;
    left: -1px;
    position: absolute;
    transition: transform 0.3s;
    background-color: ${p => p.theme.colors.white};
    box-shadow: 0 0 3px ${p => p.theme.colors.darkestContrast};
  }

  &[aria-checked='true'] {
    background-color: ${p => p.theme.colors.success};
    border-color: ${p => p.theme.colors.success};

    ::after {
      transform: translateX(26px);
    }

    ${StyledX} {
      opacity: 0;
    }

    ${StyledCheck} {
      opacity: 1;
    }
  }

  ${StyledX} {
    opacity: 1;
    transition: opacity 0.3s;
  }

  ${StyledCheck} {
    opacity: 0;
    transition: opacity 0.3s;
  }

  &[disabled] {
    opacity: 0.5;
  }

  ${margin}
`

// @ts-ignore
Toggle.propTypes = {
  value: PropTypes.bool,
  disabled: PropTypes.bool,
}

Toggle.defaultProps = {
  disabled: false,
}

export default Toggle
