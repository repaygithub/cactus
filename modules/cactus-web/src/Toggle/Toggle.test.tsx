import userEvent from '@testing-library/user-event'
import pick from 'lodash/pick'
import * as React from 'react'

import renderWithTheme from '../../tests/helpers/renderWithTheme'
import Toggle from './Toggle'

// Use this to avoid PropTypes error regarding readonly field.
const noop = () => undefined

describe('component: Toggle', () => {
  test('Should support margin space props', () => {
    const { getByTestId } = renderWithTheme(
      <Toggle checked={false} marginBottom={4} onChange={noop} data-testid="toggle" />
    )
    const toggle = getByTestId('toggle').parentElement
    const style = window.getComputedStyle(toggle as HTMLElement)

    expect(style.marginBottom).toBe('16px')
  })

  test('Should support ref prop', () => {
    const ref = React.createRef<HTMLInputElement>()
    const { getByTestId } = renderWithTheme(
      <Toggle data-testid="toggle" defaultChecked ref={ref} />
    )
    expect(getByTestId('toggle')).toBe(ref.current)
    expect(ref.current).toBeChecked()
  })

  test('Should trigger onChange event', () => {
    const box: any = {}
    const onChange = jest.fn((e) => Object.assign(box, pick(e.target, ['id', 'checked'])))
    const { container } = renderWithTheme(
      <Toggle checked={false} onChange={onChange} id="will-change" />
    )

    userEvent.click(container.querySelector('#will-change') as Element)
    expect(onChange).toHaveBeenCalled()
    expect(box).toEqual({ id: 'will-change', checked: true })
  })

  test('Should not trigger onChange event', () => {
    const onChange = jest.fn()
    const { getByTestId } = renderWithTheme(
      <Toggle disabled onChange={onChange} data-testid="will-not-change" />
    )

    userEvent.click(getByTestId('will-not-change'))
    expect(onChange).not.toHaveBeenCalled()
  })

  describe('with theme customization', () => {
    test('Should not have box shadows on focus', () => {
      const { getByTestId } = renderWithTheme(<Toggle data-testid="toggle" />, {
        boxShadows: false,
      })

      const toggle = getByTestId('toggle')
      const style = window.getComputedStyle(toggle)

      expect(style.boxShadow).toBe('')
    })
  })
})
