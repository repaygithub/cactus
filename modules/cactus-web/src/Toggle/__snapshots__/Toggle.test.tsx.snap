// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Toggle should initialize value to true 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 6px;
  right: 8px;
  color: hsl(0,0%,100%);
}

.c4 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 4px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c0 {
  position: relative;
  width: 51px;
  height: 26px;
  border-radius: 13px;
  outline: none;
  background-color: hsl(353,84%,44%);
  border: 1px solid hsl(353,84%,44%);
  cursor: pointer;
}

.c0:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c0::after {
  width: 26px;
  height: 26px;
  border-radius: 50%;
  content: '';
  top: -1px;
  left: -1px;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(200,10%,20%);
}

.c0[aria-checked='true'] {
  background-color: hsl(145,89%,28%);
  border-color: hsl(145,89%,28%);
}

.c0[aria-checked='true']::after {
  -webkit-transform: translateX(26px);
  -ms-transform: translateX(26px);
  transform: translateX(26px);
}

.c0[aria-checked='true'] .c1 {
  opacity: 0;
}

.c0[aria-checked='true'] .c3 {
  opacity: 1;
}

.c0 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0[disabled] {
  opacity: 0.5;
}

.c5[aria-checked='true'] .c1 {
  opacity: 0;
}

.c5[aria-checked='true'] .c3 {
  opacity: 1;
}

.c5 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c5 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

<button
    aria-checked="true"
    class="c0"
    role="switch"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c1 c2"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="c3 c4"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`component: Toggle should render a disabled toggle 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 6px;
  right: 8px;
  color: hsl(0,0%,100%);
}

.c4 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 4px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c5[aria-checked='true'] .c1 {
  opacity: 0;
}

.c5[aria-checked='true'] .c3 {
  opacity: 1;
}

.c5 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c5 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 {
  position: relative;
  width: 51px;
  height: 26px;
  border-radius: 13px;
  outline: none;
  background-color: hsl(353,84%,44%);
  border: 1px solid hsl(353,84%,44%);
  cursor: cursor;
}

.c0:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c0::after {
  width: 26px;
  height: 26px;
  border-radius: 50%;
  content: '';
  top: -1px;
  left: -1px;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(200,10%,20%);
}

.c0[aria-checked='true'] {
  background-color: hsl(145,89%,28%);
  border-color: hsl(145,89%,28%);
}

.c0[aria-checked='true']::after {
  -webkit-transform: translateX(26px);
  -ms-transform: translateX(26px);
  transform: translateX(26px);
}

.c0[aria-checked='true'] .c1 {
  opacity: 0;
}

.c0[aria-checked='true'] .c3 {
  opacity: 1;
}

.c0 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0[disabled] {
  opacity: 0.5;
}

<button
    aria-checked="false"
    class="c0"
    disabled=""
    role="switch"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c1 c2"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="c3 c4"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`component: Toggle should render a toggle 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 6px;
  right: 8px;
  color: hsl(0,0%,100%);
}

.c4 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 4px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c0 {
  position: relative;
  width: 51px;
  height: 26px;
  border-radius: 13px;
  outline: none;
  background-color: hsl(353,84%,44%);
  border: 1px solid hsl(353,84%,44%);
  cursor: pointer;
}

.c0:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c0::after {
  width: 26px;
  height: 26px;
  border-radius: 50%;
  content: '';
  top: -1px;
  left: -1px;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(200,10%,20%);
}

.c0[aria-checked='true'] {
  background-color: hsl(145,89%,28%);
  border-color: hsl(145,89%,28%);
}

.c0[aria-checked='true']::after {
  -webkit-transform: translateX(26px);
  -ms-transform: translateX(26px);
  transform: translateX(26px);
}

.c0[aria-checked='true'] .c1 {
  opacity: 0;
}

.c0[aria-checked='true'] .c3 {
  opacity: 1;
}

.c0 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0[disabled] {
  opacity: 0.5;
}

<button
    aria-checked="false"
    class="c0"
    role="switch"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c1 c2"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="c3 c4"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`component: Toggle should support margin space props 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 6px;
  right: 8px;
  color: hsl(0,0%,100%);
}

.c4 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 4px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c5[aria-checked='true'] .c1 {
  opacity: 0;
}

.c5[aria-checked='true'] .c3 {
  opacity: 1;
}

.c5 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c5 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6[aria-checked='true'] .c1 {
  opacity: 0;
}

.c6[aria-checked='true'] .c3 {
  opacity: 1;
}

.c6 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 {
  position: relative;
  width: 51px;
  height: 26px;
  border-radius: 13px;
  outline: none;
  background-color: hsl(353,84%,44%);
  border: 1px solid hsl(353,84%,44%);
  cursor: pointer;
  margin-bottom: 16px;
}

.c0:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c0::after {
  width: 26px;
  height: 26px;
  border-radius: 50%;
  content: '';
  top: -1px;
  left: -1px;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
  box-shadow: 0 0 3px hsl(200,10%,20%);
}

.c0[aria-checked='true'] {
  background-color: hsl(145,89%,28%);
  border-color: hsl(145,89%,28%);
}

.c0[aria-checked='true']::after {
  -webkit-transform: translateX(26px);
  -ms-transform: translateX(26px);
  transform: translateX(26px);
}

.c0[aria-checked='true'] .c1 {
  opacity: 0;
}

.c0[aria-checked='true'] .c3 {
  opacity: 1;
}

.c0 .c1 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0 .c3 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c0[disabled] {
  opacity: 0.5;
}

<button
    aria-checked="false"
    class="c0"
    role="switch"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c1 c2"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="c3 c4"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
  </button>
</DocumentFragment>
`;
